src/main/template/vpc.yaml

```diff
@@ -0,0 +1,442 @@
+      SubnetId: !Ref privateSubnetA
+
+  privateSubnetBRouteTableAssociation:
+    Type: AWS::EC2::SubnetRouteTableAssociation
+    Properties:
+      RouteTableId: !Ref privateRouteTableB
+      SubnetId: !Ref privateSubnetB
+
+  privateSubnetCRouteTableAssociation:
+    Type: AWS::EC2::SubnetRouteTableAssociation
+    Properties:
+      RouteTableId: !Ref privateRouteTableC
+      SubnetId: !Ref privateSubnetC
+
+  bastionSecurityGroup:
+    Type: AWS::EC2::SecurityGroup
+    Properties:
+      GroupName: !Sub ${AWS::StackName} Bastion Security Group
+      GroupDescription: !Sub ${AWS::StackName} Bastion Security Group
+      VpcId: !Ref vpc
+      SecurityGroupIngress:
+      - IpProtocol: tcp
+        FromPort: 22
+        ToPort: 22
+        CidrIp: 0.0.0.0/0
+      Tags:
+        - Key: Name
+          Value: !Sub ${AWS::StackName} Bastion Security Group
+
+  bastionLaunchConfiguration:
+    Type: AWS::AutoScaling::LaunchConfiguration
+    Properties:
+      ImageId: !Ref bastionAmi
+      KeyName: !Ref keyName
+      InstanceType: !Ref bastionInstanceType
+      SecurityGroups:
+        - !Ref bastionSecurityGroup
+      UserData:
+        Fn::Base64: !Sub |
+           #!/bin/bash -xe
+           yum update -y
+           /opt/aws/bin/cfn-init -v --configsets setup --stack ${AWS::StackName} --resource bastionLaunchConfiguration --region ${AWS::Region}
+           /opt/aws/bin/cfn-signal -e $? --stack ${AWS::StackName} --resource bastionAutoScalingGroup --region ${AWS::Region}
+    Metadata:
+      AWS::CloudFormation::Init:
+        configSets:
+          setup:
+            - "ssh_keys"
+            - "ntp"
+        ssh_keys:
+          files:
+            /home/ec2-user/.ssh/additional_keys:
+              source: !Ref authorizedKeys
+              owner: root
+              group: root
+              mode: 000644
+          commands:
+            register_additional_keys:
+              command: "cat /home/ec2-user/.ssh/additional_keys >> /home/ec2-user/.ssh/authorized_keys"
+        ntp:
+          packages:
+            yum:
+              ntp: []
+          services:
+            sysvinit:
+              ntpd:
+                enabled: true
+                ensureRunning: true
+                packages:
+                  yum:
+                    - ntp
+
+  bastionAutoScalingGroup:
+    Type: AWS::AutoScaling::AutoScalingGroup
+    Properties:
+      LaunchConfigurationName: !Ref bastionLaunchConfiguration
+      MaxSize: 2
+      MinSize: 0
+      DesiredCapacity: 1
+      VPCZoneIdentifier:
+        - !Ref publicSubnetA
+        - !Ref publicSubnetB
+        - !Ref publicSubnetC
+      Tags:
+        - Key: Name
+          Value: !Sub ${AWS::StackName} Bastion
+          PropagateAtLaunch: true
+    CreationPolicy:
+      ResourceSignal:
+        Timeout: PT5M
+
+Outputs:
+  vpc:
+    Description: A reference to the created VPC.
+    Value: !Ref vpc
+    Export:
+      Name: !Sub ${AWS::StackName}-vpc
```

Comment:
What's the rationale for writing `Name` tags of all resources with spaces (`MyGreatStack Private Route Table C`), but names of outputs with dashes (`MyGreatStack-public-subnets`)? Does it make sense to unify them one way or the other?

Reply:
The name tags are meant as speaking names to an AWS console user. The export names are identifiers to be used by other stacks.